### PR TITLE
ids_for_business API

### PR DESCRIPTION
**Pull request by tleekkul**
Change the way we do the facebook login, from using debug-token API to use ids-for-business API

### CODE REVIEW

File: Cpp/odin/facebook.cpp [hunk 1]
```diff
@@ -15,67 +15,125 @@
...
-    fostlib::http::user_agent ua(fb_url);
-    auto response = ua.get(fb_url);
-    auto response_data = fostlib::coerce<fostlib::string>(
-            fostlib::coerce<fostlib::utf8_string>(response->body()->data()));
-    fostlib::json body = fostlib::json::parse(response_data);
-    return fostlib::coerce<bool>(body["data"]["is_valid"]);
+    std::unique_ptr<fostlib::http::user_agent::response> get_or_mock(
+            fostlib::http::user_agent &ua,
+            fostlib::url url,
+            fostlib::json config = {}) {
```
Comment (KayEss) [inline Cpp/odin/facebook.cpp]: If you want to set a default I think you should do it in the header, not here.

File: Cpp/odin/facebook.cpp [hunk 2]
```diff
@@ -15,67 +15,125 @@
...
+            fostlib::json::parse(fostlib::coerce<fostlib::string>(
                     fostlib::coerce<fostlib::utf8_string>(
-                            response->body()->data()));
-            return fostlib::json::parse(response_data);
+                            ids_for_biz_resp->body()->data())));
+    if (!ids_for_biz.has_key("data")) {
+        fostlib::log::error(c_odin)("Error", "ids_for_business")(
+                "URL", ids_for_biz_url)("status", ids_for_biz_resp->status())(
+                "body", ids_for_biz);
+        // TODO: Should return 422
```
Comment (KayEss) [inline Cpp/odin/facebook.cpp]: Why 422?

File: Cpp/odin/facebook.cpp [hunk 3]
```diff
@@ -15,67 +15,125 @@
...
         }
+        if (std::find(allow_apps.begin(), allow_apps.end(), app_id)
+            != allow_apps.end()) {
+            // found in allow_apps
+            allowed = true;
+        }
+    }
+    if (!allowed || !fb_user.has_key("id")) {
+        // One of these two not found in ids_for_biz, raise exception
+        // TODO: Should return 422
```
Comment (KayEss) [inline Cpp/odin/facebook.cpp]: Again, why 422 here?

File: Cpp/odin/facebook.cpp (hunk 1 above)
Comment (tleekkul) [reply Cpp/odin/facebook.cpp]: I defined this function in an anonymous namespace, this also doesn't include in the header as well.

Comment (KayEss) [reply Cpp/odin/facebook.cpp]: Oh, OK. That's fine then. It does look odd though. When I first read it I thought "empty object", but of course it's the default constructor which is `null`. `fostlib::null` might be a clearer value to use

**Pull request closed without merging**